File: nattybear/level2/괄호-변환/solution.js
class Stack {
  constructor() {
    this.values = []
  }

  empty() {
    return this.values.length == 0
  }

  peek() {
    const size = this.values.length
    return this.values[size - 1]
  }

  pop() {
    this.values.pop()
  }

  push(value) {
    if (this.empty()) {
      this.values.push(value)
      return
    }
    switch (value) {
      case '(':
        this.values.push(value)
        break
      case ')':
        if (this.peek() == '(')
          this.pop()
        else
          this.values.push(value)
        break
    }
  }
}

function solution(p) {
  return to_valid(p)
}

function count(xs, c) {
  let count = 0
  for (let x of xs)
    if (x == c)
      count++
  return count
}

function balance(p) {
  const open = count(p, '(')
  const close = count(p, ')')
  return open == close
}

function valid(p) {
  const stack = new Stack()
  for (let c of p)
    stack.push(c)
  return stack.empty()
}

function split(u) {
  const v = u.slice(0, 2)
  const w = u.slice(2)
  return _split(v, w)
}

function _split(v, w) {
  if (balance(v) || w.length == 0)
    return [v, w]
  const head = v.concat(w.slice(0, 2))
  const tail = w.slice(2)
  return _split(head, tail)
}

function to_valid(w) {
  if (w.length == 0 || valid(w))
    return w

  const [u, v] = split(w)

  if (valid(u))
    return u.concat(to_valid(v))

  return ["(", to_valid(v), ")", strip_reverse(u)].join("")
}

function strip_reverse(s) {
  return s
    .slice(1, -1)
    .split("")
    .map(flip)
    .join("")
}

function flip(p) {
  if (p == "(")
    return ")"
  else
    return "("
}

File: nattybear/level2/괄호-변환/Solution.java
import java.util.ArrayList;
import java.util.Stack;

class Solution {
  public String solution(String p) {
    return toValid(p);
  }

  static int count(String s, char c) {
    int n = 0;
    for (int i = 0; i < s.length(); i++)
      if (s.charAt(i) == c)
        n++;
    return n;
  }

  static boolean balance(String p) {
    int open = count(p, '(');
    int close = count(p, ')');
    return open == close;
  }

  static boolean valid(String p) {
    CStack stack = new CStack();
    for (int i = 0; i < p.length(); i++)
      stack.puyo(p.charAt(i));
    return stack.empty();
  }

  static ArrayList<String> split(String u) {
    String v = u.substring(0, 2);
    String w = u.substring(2);
    return _split(v, w);
  }

  static ArrayList<String> _split(String v, String w) {
    if (balance(v) || w.length() == 0) {
      ArrayList<String> list = new ArrayList<>();
      list.add(v);
      list.add(w);
      return list;
    }
    String head = v.concat(w.substring(0, 2));
    String tail = w.substring(2);
    return _split(head, tail);
  }

  static String toValid(String w) {
    if (w.length() == 0 || valid(w))
      return w;

    ArrayList<String> list = split(w);
    String u = list.get(0);
    String v = list.get(1);

    if (valid(u))
      return u.concat(toValid(v));

    StringBuilder builder = new StringBuilder();
    builder.append("(");
    builder.append(toValid(v));
    builder.append(")");
    builder.append(stripFlip(u));
    return builder.toString();
  }

  static String stripFlip(String s) {
    int n = s.length();
    int[] xs = strip(s)
      .chars()
      .map(Solution::flip)
      .toArray();
    return new String(xs, 0, xs.length);
  }

  static String strip(String s) {
    int n = s.length();
    return s.substring(1, n-1);
  }

  static int flip(int p) {
    if ((char) p == '(')
      return (int) ')';
    else
      return (int) '(';
  }
}

class CStack extends Stack<Character> {
  void puyo(Character value) {
    if (empty()) {
      push(value);
      return;
    }
    switch (value) {
      case '(':
        push(value);
        break;
      case ')':
        if (peek() == '(')
          pop();
        else
          push(value);
    }
  }
}
